### ISSUE TITLE
Changing Portable version folder location:  settings file problems

### ISSUE THREAD

**Issue by tajmone** [wereturtle/ghostwriter#58]
Under Windows 10…

After moving ghostwriter portable edition's folder from one location to another, ghostwriter started quitting every time I attempted to **Preview in HTML**.

I realized that the issue had to do with the `ghostwriter.ini` settings file: after deleting it, ghostwriter's Preview in HTML worked again. Apparently, the `.ini` file didn't contain any explicity references to the older path, yet it was its presence that made the app crash.

**NOTE**: I've added MultiMarkdown's exe file inside ghostwriter's root folder (so that it reckognizes it for preview and export, even though it's not inside a folder within the %PATH% env var), and in the `.ini` file (at the time of crashing on HTML Preview) it showed up that last export operation was done in MultiMarkdown. Not sure if this might be the cause of problem (ie: maybe closing due to some error returned from `multimarkdown.exe`?)

Anyhow, whatever caused ghostwriter to abruptly quit has to do with the .ini file, it seems that ghostwriter in portable mode doesn't sanitize the saved settings to verify if it was moved since last use. This could possibly lead to failure of use in portable mode on USB sticks: on different machine the drive letter being different it might cause crash on HTML Preview.

Since the solution is to simple delete the ini file, I haven't really done  further testing to isolate the problem (moving the folder again after removing  `multimarkdown.exe`, to see if the problem is that, ecc.). But I'm willing to if needed to solve the issue.

PS: ghostwriter is simply **_wonderful_**.

**Comment by tajmone** [comment]
# ERRATA CORRIGE:
I've stumbled upon the quit on HTML preview again.
It seems it's not related to the settings file or having moved the portable folder around. It looks like it fails to Preview any doc which is not in the same folder as ghostwriter executable (which, in my case, also contains the multimarkdown exe file) when last previewing was done in multimarkdown — by default, it will open preview with last MD flavour used, so it will fail to find the MM-5 exe binary if it's not in the path of the doc being previewed. 

So, it seems that `multimarkdown.exe` has to be on the %PATH% after all... I was hoping to being able to distribute a package I'm working on in a completely stand alone format. But it looks like I'll have to find a way to attach ghostwriter's execution to a temporary env with the added path (maybe by going through a batch or PowerShell file).

I though that ghostwriter (both portable and installed version) would look for the extra markdown binaries first in its executable folder (this would allow to work with a different versions of multimarkdown from the one installed on the system). Instead, it seems that it takes the document path as the working directory where to look for the .exe file first.

**Comment by tajmone** [comment]
# ERRATA 2: Problem has to do with Metadata involving external files!

Now I've added `multimarkdown.exe` to system path, but it still quits on preview sometimes!
So it might not be either of the two point mentioned above, or maybe there is an additional problem to it.

I'm using MM-5, the latest binary release. It seems to me that the problem is in the way ghostwriters previews HTML. For some reason, some multimarkdown files fail to preview (eve though they are exported ok).

## Case 1: Metadata includes headers linking to external JS script

Specifically: the problem shows up when markdown includes metadata with headers. I was using an example from MM5 website:
```
html header:    <link rel="stylesheet" href="http://yandex.st/highlightjs/7.3/styles/default.min.css">
    <script src="http://yandex.st/highlightjs/7.3/highlight.min.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>

Testing code highlighting in MM-5.

```perl
# Demonstrate Syntax Highlighting if you link to highlight.js #
# http://softwaremaniacs.org/soft/highlight/en/
print "Hello, world!\n";
$a = 0;
while ($a < 10) {
print "$a...\n";
$a++;
}
```

which will include javascript for syntax highlighting. So I guess the HTML previewer might conflict with this.

### After all, MM-5 exe only needs to be in ghostwriter's main folder!

As a matter of fact, I've moved again the `multimarkdown.exe` file from the in-path folder back to ghostweriter portable ed. main folder, and it does reckognize and integrate MM5 export and preview in any document (even if in folders unrelated to where the exe is).

### Problem is not MM-5 related: Quits also in sundown Preview!

NOTE: **HTML Preview** will cause ghostwriter to quit abruptly on this doc with JS headers even when previewing in sundown mode (ie: when it defaults to sundown because it was last mode used). So the problem has to do with JS!

## Case 2: CSS Metadata also breaks Preview!

I further tested MM-5's metadata. It seems that Preview crashes whenever linking an outside file is envolved. For example, using the CSS metadata caused preview crash:

```
CSS:      http://example.com/standard.css
```

Note, that it exports without problem (and metadata is converted finely into HTML), only crashes on Preview!

But other metadata headers work fine, both with and without YAML style separators. For examples, this works fine in Preview:

```
---
Title:    A Sample MultiMarkdown Document  
Author:   Fletcher T. Penney  
Date:     February 9, 2011  
Comment:  This is a comment intended to demonstrate  
          metadata that spans multiple lines, yet  
          is treated as a single value.  
...
Just a test with some basic metatada.
```

I guess for the time being all I have to do is remember not to preview MM5 docs that use CSS or JS headers.

Any solution to this problem?

I apologyse for the verbosity of this issue report, and the frequent changes of heart — but I hope it helped shed some light on MM-5 integration in portable mode, apart from bringing up the issue of HTML Preview breaking up.

**Comment by wereturtle** [comment]
Well, I had never intended for ghostwriter to look for other Markdown exporter executables in its folder in portability mode.  But now that you mention it, I suppose it'd be a good idea.

Regarding the CSS and JavaScript issues with MMD, I'm afraid that's a WebKit thing.  It's out of my hands.  In the case of CSS metadata, it's possible that my forcing a stylesheet on WebKit (the default Github theme, for example) in the HTML Preview is conflicting with the CSS you set via MultiMarkdown's metadata, causing WebKit to fight with itself.  The only reasonable solution would be for me to remove the CSS feature from the HTML Preview window, which would make the entire preview rather dull.  The other way would be to preappend the HTML Preview stylesheet to the HTML being fed into WebKit, which I used to do before I found the Qt WebKit API call to set the stylesheet without modifying the HTML.  I'd rather not change back.  The new way is much better and works faster for large documents (no String appending).

If I were you, I would file this as a bug against Qt for Qt's wrapper to WebKit.  Unfortunately, it would probably be a long time before you ever saw the fix.

I'm still not sure what would cause the JavaScript crashes.  WebKit doesn't exactly have a JavaScript engine.  It should just ignore the script tags.  That may be another one you would want to file against Qt or WebKit.  In fact, a fun exercise might be to take your exported HTML and then load it with a raw WebKit installation, and see if you can replicate the crash.  If that still doesn't replicate it, then try another Qt app that has WebKit integration (a Qt-based internet browser?).

**Comment by tajmone** [comment]
I haven't worked with Qt, so I've difficoulties figuring out what might be going on behind the scenes.

First, I though it might have to do with the fact that multimarkdown usually exports as snippets, unless there are some metadata that imply creating a full webpage with headings — and CSS metadata, as well as JS in head, prompt for a full page output rather than snippet.

I'm not sure how HTML Preview in ghostwriter handles that output—ie, if it checks for the presence of <body> tags to avoid wrapping them inside another <head><body>. But it might be the case that the error might have to do with nesting of multiple <head> declarations.

Strange that the whole app shuts down, though. Preview being a different window, I thought it wouldn't have caused also the parent window to crash.

I'll have a look at the Qt documentation though, maybe it mentions if there are some problems with pages with invalid HTML (like nesting mory than one <body>, ecc).

**Comment by wereturtle** [comment]
I tried on Windows 7 to replicate the issue regarding ghostwriter not finding markdown exporter executables in its own directory, but I couldn't.  As a test, I put in the two Pandoc executables into my portable ghostwriter folder alongside ghostwriter.exe.  I then renamed the original Pandoc executables in the installation location to be different so that they could not be found on the system path.  ghostwriter was able to see the Pandoc in its own portable folder for both HTML preview and export.  It also remembered Pandoc as my last-used exporter when I closed and reopened the application.

Perhaps this is a Windows 10 thing?  I can't fix what I can't replicate, so I am going to close this issue for now.  If you ever come up with a fix, feel free to put in a pull request.

**Comment by tajmone** [comment]
You are right, I was wrong: I had interpreted the shutdown of ghostwriter HTML Preview as a symptom of not being able to find the executables in place, instead it was due to failure to visualize HTML containing CSS or JS (all the files I was using for test had some MM headers producing CSS or JavaScript).

So it has to do with the Qt previewer control, which either doesn't like CSS and JS incorporated into the page, or it might be that its expecting an HTML snippet as output (to wrap it withing added header/body opening and closing tag-sets), while the presence of CSS file metadata specifications (or JS explicit headers) will force both multimarkdown and Pandoc to produce a stand-alone document instead of a snippet. 

Maybe nesting multiple header/body tags is what brings the HTML Previer to a shutdown — in this case sanitation of processors rendered HTML, before feeding it to Qt previewer, might solve the issue.

**Comment by wereturtle** [comment]
Well, I've been trying your JavaScript/CSS snippet for a bit with both MutliMarkdown versions 4 and 5, and the HTML Preview is behaving.  Granted, I've tested under Linux.  The version of Qt 5 I am using on Linux is 5.2.1, which is a bit older than the version that ships with the Windows version of ghostwriter.  I am thinking of a few possible different things that might be happening:

1. The bug is only in the Qt version of Windows.
2. The bug is only in that version of Qt (5.5.0)
3. Some combination of 1 and 2.
4. The bug might also be happening in Linux, but it's so rare that I haven't caught it yet.

I could try bumping up the Qt version that ships with the Windows version of ghostwriter to the latest (version 5.5.6), and see if that helps the issue.  It's a long shot, though.  If you like, you could see if it helps.  Just download and install the latest version of Qt Creator with mingw32 (NOT the Visual Studio edition), and then use it to open the ghostwriter.pro file.  After setting up your build environment (where the Release exe goes etc.) you should just be able to hit the green Play button on the left.  It will compile, build, and launch it for you.

Or you can wait for me to get around to it when I'm less busy.  But I'd sure appreciate the help in determining if it's worthwhile to ship a newer version of Qt.

**Comment by tajmone** [comment]
Sorry for the delay in reply, I was away from the PC for a few days...

I somehow have the gut feelign that it's not a bug, but simply that the Win version of Qt falls under different standards of safeguarding for browser-component related contents.

I would think that the best approach is to consider what we are feeding to the HTML Previewer: what Qt is expecting is some bare HTML with not head and body tags, which is exactly what md processor would produce under most circumstances.

I think that CSS and javascript inclusion in Multimarkdown and Pandoc force them to produce a "standalone" HTML document, which is not what Qt expects. Ghostwriter should filter the output of the MD processors, and strip out all head and tails — which befits the purpose of a Preview, ie: showing only the document layout, without styling or applyed JS functions (such as code higlighting, ecc).

Also, this type of sanitation of the output, should be a security measure for malicious scripts. Surely, when it comes to browser security and behaviour, different OS behave differently.

I'll have a look at the Qt Creator you adviced me, and if there aren't any compatiblity issues with my PC I'll give it a go — just want to be sure it doesn't conflict with the current setup. I'd like to have a dev-testing ground to experiment with features and changes.

I'll let you know if I managed to set it up...

**Comment by tajmone** [comment]
Ok. I've downloaded and installed Qt Creator 5.6.0 for Win 32-bit ( MinGW 4.9.2 ).

But when I try to build it I get this error:

```
Unknown module(s) in QT: webkitwidgets
```

So it looks like I'm missing something from webkit ... Did some research, and pinned down the issue:

In the `mkspecs/modules` folder the `qt_lib_webkit.pri` and `qt_lib_webkitwidgets.pri` are not present. If my understanding is correct, these have been depracated in favour of other modules, so are no longer shipped with prebuild binaries.

From what I understand, these are deprecated modules, no longer included since Qt Creator v5.6.
I'll have to look into the Qt Creator Setup file (I used the offline installer, not the webinstaller) and see if these modules appear in the extra options during installation.

**Comment by tajmone** [comment]
Seems that Qt installare doesn't allow updating existing installation (complains about folder already existing). Mhhh a pity, I really wanted to try to add the HTML output sanitation to see if it solved the problem.

Seems I'm stuck right now. The missing webkitwidgets module issue seems to come up a lot, but you'll find mainly answers on how to fix it on Linux, not much on Windows versions (at the little I've found seems more confusing than claryfing so far).

Any ideas on how to go about it?

**Comment by tajmone** [comment]
I give up .... Uninstalled Qt Creator, reinstalled it via the web installer — and, yes: it did provide more modules, especially relating the Chromium engine — but still no trace on my HD of `QtWebKitWidgets` (no DLLs, no libraries, nothing).

It simply looks like it's no longer included. Documentation and online support for Qt is really dreadfull: it keeps pointing to the fact that adding `QT += webkitwidgets` to the `*.pro` file does the trick — but, of course, that is already there in the source where it checks for Qt version. No mentioning on how to obtain the module itself on Windows — lots of instructions for Linux though.

It looks like one has to build Qt Creator from source, on Windows, in order to have WebKit modules.

Hopefully all this might at least be helpful to update ghostwriter's REAMDE on the GH repo — as it stands, instructions for building ghostwriter on Windows don't match the problems related to Qt 5.

**Comment by wereturtle** [comment]
Wow, sorry about the trouble.  This is why I hesitate bumping up the Windows version of Qt that I distribute.  It can be a real pain when the Qt devs decide to rearrange their libraries for random and arbitrary reasons.  Or just deprecate things and forget to tell you on the release notes.  (You only find out when an algorithm stops working, because everything will build just fine.  Surprise!)

I'll have to fiddle around with it on my end a bit sometime.  I should probably preemptively check out what the changes are in 5.6, since the Linux world does tend to catch up, however slowly or quickly.

And yes, I agree, the documentation for Windows is a little sad.  But everything is just so brilliant and easy in Linux.

**Comment by wereturtle** [comment]
OK, it turns out that in Qt 5.6 they have completely dropped QWebKit, and now we're supposed to use QWebEngine instead.  They even have a [guide](https://doc-snapshots.qt.io/qt5-5.6/qtwebenginewidgets-qtwebkitportingguide.html) for porting between the two.  So, yeah, I've got my work cut out for me.

In the meantime, see if you can find Qt 5.5.  Maybe that one will help this issue?  I know ghostwriter _definitely_ compiles for 5.5, as I've used it on Linux on several VMs having that version.

(Why, oh, why, do the Qt devs have to do this on non-major versions?  A big API change should be for Qt 6, not for a 5.x release!  It violates the whole semantic version numbering system.  Grrrr.)

**Comment by tajmone** [comment]
I agree, it's a real mess depracting big thinkg in minor version updates!

But, yes: no problems, I could just install version 5.5 of Qt and give it a try with that one — after all, I'll be using Qt only for Ghostwriter, and Qt has a very clean installer/uninstaller, leaving no garbage behind.

I'll try it out with v5.5 and let you know where I get to ... — I've peeked at the code, and I have already worked out where I could try to insert sanitation code to strip the header/body tags that are added in standalone mode by the md processors. I just have to work my way through C++ syntax. Let's cross fingers!

**Comment by wereturtle** [comment]
Hi, I wanted to follow up with you concerning the porting effort of QWebKit to QtWebEngine (out of the hope that it would fix your crashing issues with JavaScript).

Right now, it seems to be a bust.  The bottom line is that the Qt devs have deprecated WebKit integration in favor of embracing Chromium, but in the process they have not yet managed to port over all the features that their WebKit integration offered.  I ran into the following issues while attempting to port ghostwriter to QtWebEngine:

- Print functionality has gone missing, which means ghostwriter will no longer be able to print to a printer from the HTML Preview window.  This functionality may be regained by using a JavaScript call, but I'm not 100% certain.
- DOM traversal functionality via a C++ API is also gone.  This is needed for inserting anchors for navigation via the Outline HUD into the HTML.  (The C++ algorithm would need to be converted to JavaScript.  I have no time for this and all the testing that goes with it.)
- Tons of other little things changed.  Some I think can be replicated with JavaScript calls.  For other things I have yet to find a one-for-one replacement.  The testing for each one will be quite time-consuming.  It was quickly becoming "death by a thousand cuts".

I am concerned that even if I overcome all these hurdles, the new Chromium integration might end up creating more issues/bugs than the current WebKit integration.

In short, I am staying with QWebKit until QtWebEngine matures.  Unfortunately, QWebKit is now deprecated as of 5.5.  It's still around until Qt 6, but it won't be updated with any fixes.  Also, as you found out when you went to compile for yourself, Qt 5.6 no longer includes precompiled QWebKit libraries.  According to the developer documentation, I would have to compile Qt myself to include QWebKit on Windows.  I am therefore freezing the Windows verison at Qt 5.5 so that I don't have to compile my own QWebKit.  This is unfortunate, since Qt 5.6 is supposed to come with further overall improvements for Windows 10.

Hopefully, in the long term, the QtWebEngine library will reach maturity so that I can smoothly transition ghostwriter to it with no loss of functionality and with minimal frustration.

**Comment by tajmone** [comment]
Ahah! That's interesting (I wish Qt documentation was more friendly to lookup, it seems to take ages to come to grips with such issues).

Most definitely, it doesn't seem worth (at least yet) to migrate to newer versions of Qt — after all, WebKit does the job finely as it is. Whatever benefits Chromium might introduce, markdown is and ramains what it is; so it can do fine without crossing the line from WebKit to Chromium.

I haven't found time to dedicate myself to the issue yet, but I'm sure that it can be fixed using Qt 5.5. I'm convinced that introducing a sanitation function that strips all beginning HTML up to `<BODY>` (indluded), and from `</BODY>` (included) to the end of document will prevent all crashes in Windows. So far I've experienced crashes only on documents which trigger "standalone" HTML output on Multimarkdown and Pandoc (ie: having metadata/frontmatter that require output document to include HTML headers, metatags and body tags, like specifying a CSS, a script, or other cases). None of these should be required for a Preview, so they can be stripped and prevent crash.

Hopefully I'll get down to give it a try when I have some due time to dig into it.

Thanks a lot for the update on the issue.
